### ISSUE TITLE
Linux and Mac Build

### ISSUE THREAD

**Issue by udnaan** [HackerPoet/NonEuclidean#1]
Since it is already written using OpenGL, would it be possible to have a build for linux and Mac?

**Comment by HackerPoet** [comment]
It may be possible, but it uses a lot of Windows APIs for other things like mouse/keyboard input, windowing, and sound/music (future).  If you're interested in porting, I can provide help, but I don't have a Mac or Linux machine for development at the moment.

**Comment by udnaan** [comment]
Fair enough. I'll look into it.

**Comment by Apostolique** [comment]
It should be possible to port this project to MonoGame. Then you can build for any current gen console along with Windows, Mac, Linux.

**Comment by necronoise** [comment]
Since linux have the glew lib, maybe this can work? Also proton/wine stuff is here to help too.

**Comment by mm12** [comment]
how would I run it??

**Comment by mm12** [comment]
> Fair enough. I'll look into it.
> It should be possible to port this project to MonoGame. Then you can build for any current gen console along with Windows, Mac, Linux.

LMK when you do I dont know how to use it

**Comment by VergeDX** [comment]
`wine NonEuclidean.exe` works OK.

**Comment by mm12** [comment]
> `wine NonEuclidean.exe` works OK.

1) love ur pic, 
2) Wine hates me so I can't. Explain what engine and winetricks to use?

**Comment by hkeena98** [comment]
> > `wine NonEuclidean.exe` works OK.
> 
> 1. love ur pic,
> 2. Wine hates me so I can't. what file should "run when I double click the application"

Yeah just running wine NonEuclidean.exe has worked out pretty well for me trying running it on anything Unix based. Why is wine not working out for you?

**Comment by KTRosenberg** [comment]
Any progress on this?
Can we get a full list of windows dependencies? If it's just the input and windowing, then there should be no issue.

Also, are all of the OpenGL features covered in WebGL as well? If so, the whole program could be ported for mobile devices on the Web to use WebVR, assuming the rendering isn't THAT expensive. At first glance, I do not see compute shaders and the like.

**Comment by ehx-v1** [comment]
Re the Wine problems, seems like Wine just wants a default executable to take care of when not called from the command line. Should be some simple initial configuration stuff, just use literally any Windows executable you use often.

**Comment by danwdart** [comment]
You could use SDL for maximal portability, I'd have thought you'd have done that since you already used a portable 3D renderer. I was going to mention I couldn't find a configure or Makefile...

**Comment by Aeyk** [comment]
I support the SDL port initiative and am interested. 

I went through and looked at the files that use Windows.h and (unsurprisingly) Engine.cpp is the biggest user.
The following places use what looks to be a Windows API. I can look into the portable equivalents later. 

** Engine.cpp
   hWnd(NULL), hDC(NULL), hRC(NULL)
   HWND, UINT, WPARAM, LPARAM 
   GetWindowLongPtr
   eng->WindowProc(hWnd, uMsg, wParam, lParam);
   DefWindowProc(hWnd, uMsg, wParam, lParam);
   SetProcessDPIAware();
   CreateGLWindow();
   InitGLObjects();
   SetupInputs();
   //Recieve events from this window
   SetWindowLongPtr(hWnd, GWLP_USERDATA, (LONG_PTR)this);
   TranslateMessage(&msg);
   DispatchMessage(&msg);
   HWND hCurWnd, UINT uMsg, WPARAM wParam, LPARAM lParam

   Engine::WindowProc
   WNDCLASSEX
   GetModuleHandle
   CreateWindowEx
   DescribePixelFormat
   SetPixelFormat
   MessageBoxEx
   ChoosePixelFormat
   RAWINPUTDEVICE
   RegisterRawInputDevices
   GetWindowRect
   SetCursorPos
   GetSystemMetrics

** Input.cpp
   Memset is portable, correct?
   GH_MOUSE_SMOOTH
   raw->header.dwType
   RI_MOUSE_RIGHT_BUTTON_DOWN
   as some examples, these seem like they would simple enough in SDL.
** Timer.h
   This has only one reference to Win32 API, and it's about timing.
   QueryPerformanceCounter
** Player.cpp
   (Nothing?)

**Comment by ehx-v1** [comment]
To clarify my previous comment:

> 2. Wine hates me so I can't. what file should "run when I double click the application"

This sounds like Wine wants to know what the "default application" for when not starting Wine from commandline is. Set it up with some .exe you use often. (I don't have experience with Wine, but this seems like the only thing that makes sense.)

**Comment by udnaan** [comment]
To anyone trying to run it through wine, try [lutris](https://lutris.net/).

Since the time this issue was opened (2018), things have progressed significantly on Linux gaming front and I bet that there is an existing lutris profile that will run the window exe as it is. Disclaimer: I have not tried it.
